### ISSUE TITLE
define instruction from easy 6502 not working

### ISSUE THREAD

**Issue by lerugray** [skilldrick/6502js#7]
Assembler doesn't recognize the same instruction from the tutorial

**Comment by BigEd** [comment]
Which assembler are you trying to use? Each assembler does have unique choices for how to express a program, and easy6502 is no exception!

**Comment by lerugray** [comment]
I tried downloading the 6502 JS Assembler, it would not recognize the
Define instruction as used in the tutorials.

**Comment by BigEd** [comment]
Do you mean http://www.6502asm.com/ ?  If so, it has no define, so from the examples we see the tactic is to use a label:
```
* = $hhhh
name
```

**Comment by BigEd** [comment]
Oops sorry I have just realised that I was confused between easy6502 and 6502js.  Let me look into this. We do know the two assemblers have diverged. My own fork of 6502js is different again.

**Comment by lerugray** [comment]
OK Cool thank you, yeah I was hoping that there was one of these I could
download to practice assembly on my own, like building snake from scratch
with the embedded screen you guys had in easy 6502

if this is possible, plz lemme know, I know easy6502 had a define
instruction for their snake example

**Comment by lerugray** [comment]
for example, trying to use the easy 6502 assembler downloaded from easy 6502

in the tutorial, they use (define) as a means of declaring
constants/variables. Upon downloading the Assembler though, the define
instruction stops working. How would I still define variables/constants?

**Comment by BigEd** [comment]
Just started looking at this.  I see you can use the same kind of workaround, abusing the label mechanism to define a symbol:

```
* = 255
mask:

* = $600
start:
nop
lda mask
nop
```

**Comment by lerugray** [comment]
How would you use this technique in something like a constant? Lets say for
example I wanted to set sysRandom = $fe (the RNG on the mock 6502)

*=255 - is this setting mask to 255 decimal? I'm a little confused as to
how this example works in practice, thanks for the help

**Comment by BigEd** [comment]
Here's a slight expansion of the example:

```
* = 127
mask:

* = $fe
sysRandom:

* = $600
start:
nop
lda mask
and sysRandom
nop
```

and the disassembler says we got this:
```
Address  Hexdump   Dissassembly
-------------------------------
$0600    ea        NOP 
$0601    a5 7f     LDA $7f
$0603    25 fe     AND $fe
$0605    ea        NOP 
```

So, it's a little ugly, and it takes two lines to make each define, but I think it works generally.

**Comment by BigEd** [comment]
(However, I notice the tactic fails for immediate values. I should have written
lda #mask 
but that turns out not to work at all.)

**Comment by lerugray** [comment]
Aghh that's unfortunate,  was going to use this idea for costants related
to collision, if there's a workaround,  kindly let me know, thx!

**Comment by lerugray** [comment]
oh and just for context, I'm trying to build snake/pong in the assembler
without using the help of the tutorial, so i assume even if reformatted for
the above hack, that code still wouldnt assemble given the #immediatevalue
error.

**Comment by BigEd** [comment]
I think we'd need to bring the whole 'define' mechanism in from easy6502 to 6502js.

Meantime, could you work using a local copy of easy6502 instead of using 6502js? I tend to try out small snippets just using the first emulation window - the snake is way down the file and needn't distract you.

**Comment by lerugray** [comment]
Yes that would be perfect, could you link me to the appropriate place to
grab the easy 6502 engine? Tha k you so much

**Comment by lerugray** [comment]
Oh wait I see what you mean, yeah that sounds good, please keep me posted
as to when you add in define. Also curious, how do I reserve bytes for
variables? For the nes, all I usually will do is (nesasm)

ballspeedx:              .rs 1

**Comment by BigEd** [comment]
Reserving some bytes... I think the effect will be to set the label and then for the next label to take a slightly higher value. So perhaps just use dummy code sequences instead:

```
* = $80
onebyte:
nop
twobytes:
nop
nop
twomorebytes:
lda 0
eightbytes:
jmp dummy
jmp dummy
lda 0
```

**Comment by codemaster138** [comment]
I made a fork of the emulator [here](https://github.com/codemaster138/6502js) and fixed the issue partially. the following would now work : 
```
define white $01
define topleft $0200
lda #white
sta topleft
```
I haven't yet managed to make it work in indexing modes though, so
```
sta $0200,#value
```
would not assemble, assuming value is a constant defined by the ```define``` keyword
